drivers/tee/tee_core.c

```diff
@@ -193,6 +199,67 @@ static int params_from_user(struct tee_context *ctx, struct tee_param *params,
 			params[n].u.memref.size = ip.u.memref.size;
 			params[n].u.memref.shm = shm;
 			break;
+
+		case TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_SECURE:
+		{
+			/*
+			 * Memref is a dmabuf file handle.
+			 * Create a struct shm that reflects the buffer
+			 * referred by the dmabuf handle.
+			 *
+			 * By calling dma_buf_map_attachment(), and since
+			 * the buffer was allocated through SMAF, tee driver
+			 * notifies a "teeX" device access request to
+			 * buffer. SMAF will eventually invoke the SMAF TA to
+			 * grant TEE to access to memory reference.
+			 *
+			 * TEE driver releases memref resources once back
+			 * from the invocation.
+			 *
+			 * TODO: should SMAF request from "teeN" device
+			 * or from 'optee0' device?
+			 *
+			 * TODO:
+			 * Use a struct shm/field flags = DMA_BUF, not MAPPED.
+			 * This tell field 'kaddr' carries reference resources,
+			 * not the kenrel virtual address.
+			 */
+			struct dmabuf_ref *ref;
+
+			if ((int)ip.u.memref.shm_id < 0)
+				return -EINVAL;
+
+			shm = kzalloc(sizeof(*shm) + sizeof(*ref), GFP_KERNEL);
+			if (!shm)
+				return -ENOMEM;
+			ref = (struct dmabuf_ref *)(shm + 1);
+
+			shm->ctx = ctx;
+			shm->teedev = ctx->teedev;
+			shm->flags = TEE_SHM_SECURE;
+			shm->kaddr = (void *)ref;
+
+			params[n].u.memref.shm_offs = ip.u.memref.shm_offs;
+			params[n].u.memref.size = ip.u.memref.size;
+			params[n].u.memref.shm = shm;
+
+			shm->dmabuf = dma_buf_get((int)ip.u.memref.shm_id);
+			if (!shm->dmabuf)
+				return -EINVAL;
+
+			ref->attach = dma_buf_attach(shm->dmabuf,
+						     &shm->teedev->dev);
```

Comment:
We could insert here `dma_buf_put(shm->dmabuf);` to release dma_buf file descriptor.
it's no more needed once we have attached to the dma_buf resource.